### PR TITLE
Checks for use of external field and solvents

### PR DESCRIPTION
**Pull request by bhourahine**
As there is a missing electrostatic term for extra fields from the solvent bulk at the moment.
- [x] stop external fields for program and API if solvent modifies the electrostatic environment
- [x] stop external polarizable media calculations if solvent modifies the electrostatic environment
- [x]  print description/warnings about dipole an/or electrostatic potentials in the presence of field modifying solvents
- [x] Block time propagation being used in cases modified by the electrostatic environment of the solvent

### CODE REVIEW

File: prog/dftb+/api/mm/mmapi.F90 [hunk 1]
```diff
@@ -387,6 +387,11 @@ subroutine TDftbPlus_setExternalPotential(this, atomPot, potGrad)
 
     call this%checkInit()
 
+    if (allocated(this%main%solvation)) then
+      write(stdOut, "(A)") "Error: External fields currently unsupported for solvent models"
+      stop
```
Comment (awvwgk) [inline prog/dftb+/api/mm/mmapi.F90:392]: Should this really use `stop` here?

File: prog/dftb+/api/mm/mmapi.F90 [hunk 2]
```diff
@@ -409,6 +414,11 @@ subroutine TDftbPlus_setExternalCharges(this, chargeCoords, chargeQs, blurWidths
...
```
Comment (awvwgk) [inline prog/dftb+/api/mm/mmapi.F90:419]: Same here, should this be a `stop`?

File: prog/dftb+/lib_dftbplus/initprogram.F90 [hunk 3]
```diff
@@ -2100,6 +2100,13 @@ subroutine initProgramVariables(this, input, env)
           & spin orbit calculations")
     end if
 
+    if (allocated(this%solvation)) then
+      if (this%tExtChrg .or. this%tEField) then
+        write(*,*)'External fields are not currently compatible with implicit solvents.'
```
Comment (awvwgk) [inline prog/dftb+/lib_dftbplus/initprogram.F90:2105]: There is no action associated with this write statement

**Review by awvwgk — CHANGES_REQUESTED** [review]
I don't think we should block solvation and external fields. At least for Born models no additional terms arise here. The check should be more concrete and only block COSMO with external fields.

For solvation and external charges I don't see a good way to put them together at all. Especially because the cavity generation for an external charge is not clear. This is something we should create at least a warning for.

**Comment by bhourahine** [comment]
> At least for Born models no additional terms arise here

Is that a correct assertion? For a spherical cavity inside a dielectric of eps_r and where there is an external homogeneous field E0, the electric field inside the 'bubble' is E0 * 3eps_r/(2eps_r +1) from Gauss' law. This should apply to a Born approximation equally as with a PCM, or am I missing something special about generalised Born? The external fields in the code are definitely not adapted for this.

File: prog/dftb+/lib_dftbplus/initprogram.F90 (hunk 3 above)
Comment (bhourahine) [reply prog/dftb+/lib_dftbplus/initprogram.F90]: Yes, that was a stupid typo on my part...

**Comment by awvwgk** [comment]
Oh right, I was only looking at the solute and cavity terms here.

**Review by aradi — APPROVED** [review]
It looks OK to me.

File: prog/dftb+/api/mm/mmapi.F90 [hunk 4]
```diff
@@ -391,6 +391,13 @@ subroutine TDftbPlus_setExternalPotential(this, atomPot, potGrad)
 
     call this%checkInit()
 
+    if (allocated(this%main%solvation)) then
+      if (this%main%solvation%isEFieldModified()) then
+        write(stdOut, "(A)") "Error: External fields currently unsupported for this solvent model"
+        stop
```
Comment (awvwgk) [inline prog/dftb+/api/mm/mmapi.F90:397]: There is still the question if we really should stop here? I don't think it is a good idea to stop in the API and return with exit code 0. If we really have to kill the API user, can we at least return with a non-zero exit code, please?

**Review by awvwgk — CHANGES_REQUESTED** [review]

File: prog/dftb+/api/mm/mmapi.F90 (hunk 4 above)
Comment (bhourahine) [reply prog/dftb+/api/mm/mmapi.F90:397]: @awvwgk are you asking for just an "error stop", or should we discuss meaningful error codes?
This brings up the need to have proper error code return from the API.

Comment (awvwgk) [reply prog/dftb+/api/mm/mmapi.F90:397]: Exactly, I'm just asking for some non-zero exit code on an error, `error stop` will do.

**Review by awvwgk — APPROVED** [review]

**Pull request merged**